Event Timing: dispatch to PerformanceObserver

francovs opened the pull request:
#### 955dc5016c8367877ebc3a3ada6f1827c3a69064
<pre>
Event Timing: dispatch to PerformanceObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=297010">https://bugs.webkit.org/show_bug.cgi?id=297010</a>

Reviewed by Ryan Reno and Ryosuke Niwa.

Implements the dispatch of PerformanceEventTiming entries to
PerformanceObsevers.

Some fixes were made in the event timing implementation:
PerformanceEventTiming entries now use WeakPtr to avoid a reference
cycle, and the use of maplike on EventCounts was replaced by a manual
implementation of .get() and .size.

Some code was moved around to better manage dependencies, which
required some header changes: PerformanceEventTimingCandidate is now
PerformanceEventTiming::Candidate, and this required PerformanceEntry.h
to no longer include Performance.h in order to avoid a circular
dependency.

* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.any-expected.txt: now fails - missing maplike interface
* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.window-expected.txt: now fails - missing maplike interface
* Source/WebCore/Headers.cmake: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/page/PerformanceMark.h: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/page/PerformanceNavigationTiming.cpp: fix build (removal of Performance.h from PerformanceEntry.h)

Canonical link: <a href="https://commits.webkit.org/298573@main">https://commits.webkit.org/298573@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f94691e32b6555d471400da1728a65f4f31a737e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115693 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121740 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43939 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87883 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42526 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21939 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65415 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98134 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22060 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96638 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42955 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100013 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96426 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24574 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41685 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19543 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38497 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42479 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48051 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41952 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45283 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43660 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -281,6 +289,29 @@ void Performance::countEvent(EventType type)
+    if (duration < 16.0)
+        return;
+
+    // We could do an even better of returning early to avoid the creation of
```
even better job? I don't think this comment is necessary.
If anything, we should add FIXME instead:
```
// FIXME: Implement a more precise early return.
```

rniwa commented on Source/WebCore/page/PerformanceObserver.h:
```diff
@@ -84,6 +85,7 @@ class PerformanceObserver : public RefCounted<PerformanceObserver> {
     bool m_registered { false };
     bool m_isTypeObserver { false };
     bool m_hasNavigationTiming { false };
+    double m_durationThreshold = 100; // In milliseconds
```
Please use `Seconds` instead of strong raw double.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2489,13 +2491,13 @@ void LocalDOMWindow::dispatchPendingEventTimingEntries()
-        // FIXME: dispatch to relevant performance observers
-        // FIXME: first-input handling
-        // FIXME: if duration > 104ms and buffer is not full, add to buffer
+        auto duration = renderingTimestamp - candidateEntry.startTime;
```
Instead of doing raw double computation, we should be using `Seconds`.
e.g. Use `performance().nowInReducedResolutionSeconds()` instead of `performance().now()`.

rniwa commented on Source/WebCore/page/PerformanceEventTiming.h:
```diff
@@ -26,35 +26,45 @@
+    struct Candidate {
+        EventTypeInfo typeInfo;
+        bool cancelable { false };
+        DOMHighResTimeStamp startTime { 0 };
```
We should use `Seconds` to store these values.

rniwa commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -127,10 +129,14 @@ Seconds Performance::timeResolution()
 
+Seconds Performance::relativeTimeFromTimeOriginInReducedResolutionSeconds(MonotonicTime timestamp) const
+{
+    return reduceTimeResolution(timestamp - m_timeOrigin);
```
Oh, I think we need to reduce the resolution before subtraction.
Otherwise, we can end up leaking a higher precision time.

francovs replied:
This implementation preserves the current `performance.now()` behavior, so I feel it would be better to do such a change separately.

rniwa commented on Source/WebCore/page/PerformanceEventTiming.h:
```diff
@@ -26,35 +26,49 @@
+    struct Candidate {
+        EventTypeInfo typeInfo;
+        bool cancelable { false };
+        Seconds startTime { 0 };
```
No need to initialize `Seconds`.

francovs replied:
I'm being forced to initialize everything because [some builds](https://github.com/WebKit/WebKit/pull/48910#issuecomment-3172987322) are failing with `-Werror=missing-field-initializers`

rreno replied:
@basuke was running into this recently too. Linux builds failing due to a missing initializer.

rniwa commented on Source/WebCore/page/PerformanceEventTiming.h:
```diff
@@ -26,35 +26,49 @@
+        Seconds startTime { 0 };
+        Seconds processingStart { 0 };
+        Seconds processingEnd { 0 };
+        RefPtr<EventTarget> target { nullptr };
```
Ditto. `RefPtr` gets initialized automatically.

rniwa commented on Source/WebCore/page/Performance.h:
```diff
@@ -153,11 +156,22 @@ class Performance final : public RefCounted<Performance>, public ContextDestruct
+
+    // Constants to avoid the need to round to PerformanceEventTiming::durationResolution
+    // when filtering candidate entries:
+    static constexpr Seconds minDurationCutoffBeforeRounding = PerformanceEventTiming::minimumDurationThreshold - (PerformanceEventTiming::durationResolution/2);
```
Nit: Please add a space around `/`.

rniwa commented on Source/WebCore/page/Performance.h:
```diff
@@ -153,11 +156,22 @@ class Performance final : public RefCounted<Performance>, public ContextDestruct
+    // Constants to avoid the need to round to PerformanceEventTiming::durationResolution
+    // when filtering candidate entries:
+    static constexpr Seconds minDurationCutoffBeforeRounding = PerformanceEventTiming::minimumDurationThreshold - (PerformanceEventTiming::durationResolution/2);
+    static constexpr Seconds defaultDurationCutoffBeforeRounding = PerformanceEventTiming::defaultDurationThreshold - (PerformanceEventTiming::durationResolution/2);
```
Ditto.

francovs committed:
Event Timing: dispatch to PerformanceObserver
https://bugs.webkit.org/show_bug.cgi?id=297010

Reviewed by Ryan Reno and Ryosuke Niwa.

Implements the dispatch of PerformanceEventTiming entries to
PerformanceObsevers.

Some fixes were made in the event timing implementation:
PerformanceEventTiming entries now use WeakPtr to avoid a reference
cycle, and the use of maplike on EventCounts was replaced by a manual
implementation of .get() and .size.

Some code was moved around to better manage dependencies, which
required some header changes: PerformanceEventTimingCandidate is now
PerformanceEventTiming::Candidate, and this required PerformanceEntry.h
to no longer include Performance.h in order to avoid a circular
dependency.

* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.any-expected.txt: now fails - missing maplike interface
* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.window-expected.txt: now fails - missing maplike interface
* Source/WebCore/Headers.cmake: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/page/PerformanceMark.h: fix build (removal of Performance.h from PerformanceEntry.h)
* Source/WebCore/page/PerformanceNavigationTiming.cpp: fix build (removal of Performance.h from PerformanceEntry.h)

Canonical link: https://commits.webkit.org/298573@main
